### ISSUE TITLE
[SYCL] Failed to find cpu

### ISSUE THREAD

**Issue by cheng-hsiang-chiu** [intel/llvm#7588]
**Describe the bug**
sycl::platform::get_platforms and platform.get_devices() fails to find the cpu, only finds the gpu information. 

**To Reproduce**
git clone https://github.com/intel/llvm -b sycl
python llvm/buildbot/configure.py --cuda
python llvm/buildbot/compile.py 

1. Specify the code
std::vector<sycl::platform> platforms = sycl::platform::get_platforms();
for (const auto& platform : platforms) {
    std::cout << platform.get_info<sycl::info::platform::name>() << '\n'
    std::vector<sycl::device> devices = platform.get_devices();
    for (const auto& device : devices) {
        std::cout << device.get_info<sycl::info::device::name>() << '\n'
        auto queue = sycl::queue(device);
        queue.submit([](sycl::handler& handler){
            handler.single_task([](){});
        });
    }
}

2. Specify the command which should be used to compile the program
clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda -std=c++17 sycl_device.cpp

3. Indicate what is wrong and what was expected
It only prints out the GPU device. 
It was expected to print out both CPU and GPU device. 

**Environment (please complete the following information):**

- OS: CentOS 7
- Target device and vendor: Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz and NVIDIA GeForce RTX 2080 Ti
- DPC++ version: clang 16.0.0
- Dependencies version: CUDA 11.6

**Additional context**

**Comment by AlexeySachkov** [comment]
Hi @cheng-hsiang-chiu,

Have you completed installation steps for OpenCL CPU RT? They are documented [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#use-dpc-toolchain)

**Comment by cheng-hsiang-chiu** [comment]
Hi @AlexeySachkov,

Thank you for the info. I used oclcpuexp-2022.14.8.0.04_rel.tar.gz and finished step 1 to 5 from the link [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#use-dpc-toolchain). I still can not get the CPU information.

**Comment by AlexeySachkov** [comment]
@cheng-hsiang-chiu,

My current guess is that something is still wrong with the environment setup you have, so a couple of steps to check that:

Could you please launch your application under `SYCL_PI_TRACE=-1` environment variable and share the beginning of that log file? The first thing I want to check is that `libpi_opencl.so` (OpenCL PI Plugin) is successfully loaded.

Could you please also attach the output gathered by launching the app under `LD_DEBUG=libs` environment variable? Looking at those logs it should be possible to pinpoint the library which wasn't loaded and therefore find out the reason why CPU device is not available.

**Comment by AlexeySachkov** [comment]
I took a look at your original post once again and I might have another guess about what happens:

```
std::vectorsycl::platform platforms = sycl::platform::get_platforms();
for (const auto& platform : platforms) {
std::cout << platform.get_infosycl::info::platform::name() << '\n'
std::vectorsycl::device devices = platform.get_devices();
for (const auto& device : devices) {
std::cout << device.get_infosycl::info::device::name() << '\n'
auto queue = sycl::queue(device);
queue.submit([](sycl::handler& handler){
handler.single_task({});
});
}
}
```

Here you try to launch a kernel on every device you discover

```
clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda -std=c++17 sycl_device.cpp
```

But this command only instructs to build and bundle device code for CUDA target, it won't include any device code for other devices into the app. Not being able to find a kernel should result in an exception that kernel wasn't found, but I don't remember exactly whether it should be synchronous or not. Async exceptions were silently ignored until we implemented a default async handler in cd93d8f64b54968f4654fcbcceab50b663220a03.

To check that second guess, you should either ensure that your toolchain version has the aforementioned commit or simply pass an async handler to `queue` constructor. You can find some example and info about async exceptions in [4.13. Error handling](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#error-handling)

**Comment by cheng-hsiang-chiu** [comment]
@AlexeySachkov 

I attach two log files. One with LD_DEBUG=libs on and the other with SYCL_PI_TRACE=-1 on. 
[SYCL_PI_TRACE.log](https://github.com/intel/llvm/files/10144811/SYCL_PI_TRACE.log)
[LD_DEBUG.log](https://github.com/intel/llvm/files/10144812/LD_DEBUG.log)

**Comment by AlexeySachkov** [comment]
From `LD_DEBUG.log`:

```
     99529:	/lib64/libstdc++.so.6: error: version lookup error: version `GLIBCXX_3.4.20' not found (required by /opt/intel/oclcpuexp-2022.14.8.0.04/x64/libintelocl.so) (fatal)
```

This is likely the reason why CPU device can't be seen.

**Comment by cheng-hsiang-chiu** [comment]
@AlexeySachkov 

I updated the /lib64/libstdc++.so.6 to a new one and re-generated the log files.
This time I got the error message, 
/opt/intel/oclcpuexp-2022.14.8.0.04/x64/libclang_compiler.so.2022.14.8.0: error: symbol lookup error: undefined symbol: RegisterGlobalAtExitNotification (fatal)
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  Native API failed. Native API returns: -42 (PI_ERROR_INVALID_BINARY) -42 (PI_ERROR_INVALID_BINARY)

I also attach the two log files for reference. 
[sycl_device.log](https://github.com/intel/llvm/files/10168559/sycl_device.log)
[LD_DEBUG.log](https://github.com/intel/llvm/files/10168560/LD_DEBUG.log)

**Comment by AlexeySachkov** [comment]
@cheng-hsiang-chiu,

> This time I got the error message,
/opt/intel/oclcpuexp-2022.14.8.0.04/x64/libclang_compiler.so.2022.14.8.0: error: symbol lookup error: undefined symbol: RegisterGlobalAtExitNotification (fatal)

That can likely be disregarded, it is not a root cause of the new problem you see.

> terminate called after throwing an instance of 'sycl::_V1::runtime_error'
what(): Native API failed. Native API returns: -42 (PI_ERROR_INVALID_BINARY) -42 (PI_ERROR_INVALID_BINARY)

This is expected based on the command line you shared in the original post. See my [previous comment](https://github.com/intel/llvm/issues/7588#issuecomment-1335894605): you compile device code only for CUDA target, but you try to launch it on every device. The exception you see is expected, because you try to unconditionally launch the kernel on every device you have. You should add `spir64` to the list of your targets to also emit device code in device-agnostic SPIR-V format so it works on non-CUDA devices as well.

I also suppose that you now see the print about CPU device, right?

**Comment by cheng-hsiang-chiu** [comment]
@AlexeySachkov 

You are right. I build the executable using the following command and it works now,
clang++ -fsycl -fsycl-targets=spir64_x86_64,nvptx64-nvidia-cuda ./sycl_device.cpp -o sycl_device

Thank you so much for your helps.
